### ISSUE TITLE
Extract libappimage from AppImageKit, making AppImageKit the home of the dev tools only

### ISSUE THREAD

**Issue by TheAssassin** [AppImage/AppImageKit#840]
In order to reduce confusion about responsibilities of certain repositories, we've already extracted appimaged from this repository. We should now do the same with libappimage.

libappimage is the "core library" of almost all AppImage related software. It provides interfaces for handling AppImages.

Extracting has a couple of advantages:

  - we could set up some "real" versioning for libappimage which is important to _many_ projects, while maintaining our continuous deployment scheme for the dev tools like appimagetool (which aren't supposed to be shipped in distros anyway)
  - appimagetool could continue to link to patched versions of libraries, whereas libappimage could be linked to the non-patched distribution versions https://github.com/AppImage/AppImageKit/issues/815
  - libappimage related tools could finally get into Debian and other more restrictive distributions (really important for AppImageLauncher, and maybe even appimaged at some point)
  - we could improve our API and finally decouple libappimage _properly_ from appimaged and other tools (this requires some work, as we'd have to redesign big portions of the current interface, but it's really worth the effort)

The only disadvantage is that it'd introduce some very small maintenance overhead for appimagetool. But as appimaged showed, this shouldn't become a major issue.

**Comment by probonopd** [comment]
Please. No. Half the Internet ;-) is linking to https://github.com/AppImage/AppImageKit/ as the home of `appimagetool` and `appimaged` (the latter of which we already moved away which created work without a clear benefit, at least for me). This type of changes really breaks things and creates work which takes the fun out of open source for me. Let's keep it simple. For the first 10 years or so _all_ of AppImage* was just 1 svn (later git) repo.

**Comment by TheAssassin** [comment]
This is the only viable solution in order to actually reduce all the maintenance overheads the old structure introduced, as it's the only clean solution to continue the development of AppImage related tools. Also, it allows us to finally improve AppImage support in distributions and desktop environments, which is something users have requested in these 10 years you mentioned.

We've come a long way to get to where we are at the moment, and the only way we could proceed is to finally create clean interfaces for working with AppImages that distributions etc. can implement. We have all of this available in the source code, actually. *Nobody* wants to write that themselves, only if *we* provide it, we will have a chance to convince DEs, app stores, etc. of implementing AppImage support.

I don't see what kind of work this would create for you. We don't expect you to code these changes. In fact, we want to contribute those things. We're slowly bringing some structure into the AppImage related software projects.

I really want to avoid this, but if you insist, then you leave us no choice than having to fork AppImageKit and maintain libappimage "externally" in the future to provide something people can use and rely on. We rely on libraries with specific versions, too, and third parties will rely on versioning when making use of our software. In the end, versioning is necessary for some types of software, especially libraries. This is what I tried to point out: there is no need to introduce versioning in AppImageKit itself if we extract the part that _requires_ versioning.

The only structural changes we would have to make:
  - make appimaged use some sort of "libappimage" repository instead of AppImageKit
  - make AppImageKit link to the libappimage repository (i.e., add a submodule to libappimage and load it from CMake, the rest of the CMake code doesn't need to be touched)

We'd be able to move a good portion of the CMake scripts into the new repository. Maintenance efforts _will_ become lower, I'm not kidding. We have to maintain smaller amounts of code at once, we can add tests to test everything separately, and the less source code has to be managed at a time, the easier it is to find specific parts inside it, and our issue trackers are easier to maintain.

Also, the same argument we had when thinking about extracting appimaged applies: a "libappimage" repository describes its content in its name. "AppImageKit" does not, it just creates confusion.

The end goal is to decouple all the components of AppImageKit from each other, which will only speed up our development, as it makes debugging etc. a lot easier. And if distributions start implementing our APIs, we'll hopefully see better AppImage support everywhere. I thought that is something especially you should appreciate.

**Comment by TheAssassin** [comment]
Re.

> And I don't want libappimage as something versioned that needs to be released, needs to be backward compatible, ABI stable, etc. Even more work. It should always be linked statically into whatever uses it. I don't have the capacity to deal with "proper distro packaging". Which is why I started AppImage in the first place ;-)

I don't feel like doing that either, and you know I'm a big fan of continuous releases. However, I can see that third parties need some sort of "identifier" they can rely on.
We just need to find a sort of "versioning scheme" that works for us. There _are_ libraries which don't do "versioning" in the sense of semver, and are in e.g., Debian, and used by third parties.

**Comment by azubieta** [comment]
@probonopd I'm going to forward to @TheAssassin in this request. Decoupling AppImageKit into modules will indeed easy development and maintenance work, as we will have to deal with small and well focused components and not whit a big and a bit messy project. Also will allow us to write more test which is good to improve the quality.
Versioning and ABI compativility is something that we will have to deal whit if we whant that, by example, the dolphin file manager shows datailed information (Description, Author, etc) about an AppImage file. 
 
About the work, don't worry we will handle it :smiley: 
Also we will find the way of making the transition as easy as possible to our current users.

**Comment by probonopd** [comment]
Yes, I can see that this could have benefits, I just won't be able to sustain this in my very limited spare time. As long as  I was basically the only contributor I would never have recommended anyone to use any of my code, but would have recommended e.g. the Dolphin team to read the AppImage specs and go from there. 

But hey, maybe the times have changed and the project has grown up.
We need to keep in mind the long-term sustainability of such changes though.

Would it be an option to move libappimage into its own repo, and leave appimagetool in AppImageKit? I think this would at least reduce the pain.

Is there a way to enforce people to always statically link libappimage and disable (or make very hard) the option to use it as a dynamic library?

**Comment by TheAssassin** [comment]
> Would it be an option to move libappimage into its own repo, and leave appimagetool in AppImageKit? I think this would at least reduce the pain.

That was the plan from the beginning. Extract only libappimage, and link to it from AppImageKit. AppImageKit then becomes the *de facto* home of all dev tools, libappimage is what implements the spec and provides convenience functions, appimaged is a separate tool.

> Is there a way to enforce people to always statically link libappimage and disable (or make very hard) the option to use it as a dynamic library?

That's a horrible idea. We're pushing to get user space tools to use libappimage. AppImageLauncher is one of them, but also Dolphin or Discover from KDE for instance could in the future use it. These tools are shipped by distributions, and distros like shared libraries. They'd just patch it away anyway, I guess. And we can still link it statically in our official tools.

The point is, nobody wants to use appimagetool's code, but we all are interested in the core code. By extracting it, we can relax some of the strict rules you put in place for the other code, and introduce some rudimentary versioning etc. But we only have to do it _there_. That's why I want to extract it.

**Comment by TheAssassin** [comment]
After many hours of fiddling with the code base, I managed to extract libappimage into a separate project: https://github.com/AppImage/libappimage.

It was a real challenge to get to this point, due to our highly complex build scripts and also way too tightly coupled code. I fixed this by splitting up the code base into three reasonable modules which are loosely coupled, which should always be the preferred way to design code bases. The unit tests also make sense now.

I already prepared a patch for AppImageKit which will make us use the libappimage repository instead of what we have currently.

There's a few TODOs left. For instance, I started to implement a hashlib, providing MD5 at the moment (needed for the AppImage runtime). Ideally, we will be able to OpenSSL for the purpose eventually.
